crates/derive/src/stages/channel_reader.rs

```diff
@@ -148,16 +157,38 @@ impl BatchReader {
     /// Pulls out the next batch from the reader.
     pub(crate) fn next_batch(&mut self, cfg: &RollupConfig) -> Option<Batch> {
         // If the data is not already decompressed, decompress it.
+        let mut brotli_used = false;
         if let Some(data) = self.data.take() {
-            let decompressed_data = decompress_to_vec_zlib(&data).ok()?;
-            self.decompressed = decompressed_data;
+            // Peek at the data to determine the compression type.
+            if data.is_empty() {
+                tracing::warn!(target: "batch-reader", "Data is too short to determine compression type, skipping batch");
+                return None;
+            }
+            let compression_type = data[0];
+            if (compression_type & 0x0F) == ZLIB_DEFLATE_COMPRESSION_METHOD ||
+                (compression_type & 0x0F) == ZLIB_RESERVED_COMPRESSION_METHOD
+            {
+                self.decompressed = decompress_to_vec_zlib(&data).ok()?;
+            } else if compression_type == CHANNEL_VERSION_BROTLI {
+                brotli_used = true;
+                self.decompressed = decompress_brotli(&data).ok()?;
+            } else {
+                tracing::warn!(target: "batch-reader", "Unsupported compression type: {:x}, skipping batch", compression_type);
```

Comment:
I imagine we don't throw a hard error here because the pipeline can't halt? This seems like something that should be more critical.

Reply:
Will error - this is a good point. Invalid batches should be skipped so we do not want to halt the pipeline here.